std/experimental/allocator/building_blocks/bitmapped_block.d

```diff
@@ -1746,3 +1760,350 @@ private struct BitVector
+            {
+                // Overestimated
+                continue;
+            }
+            _control = (cast(shared(ulong*)) data.ptr)[0 .. controlWords];
+            _control[] = 0;
+            _payload = cast(typeof(_payload)) payload;
+            break;
+        }
+        _blocks = cast(typeof(_blocks)) localBlocks;
+    }
+
+    /// Ditto
+    static if (chooseAtRuntime == theBlockSize)
+    this(ubyte[] data, uint blockSize)
+    {
+        this._blockSize = blockSize;
+        this(data);
+    }
+
+    /// Ditto
+    static if (!is(ParentAllocator == NullAllocator))
+    this(size_t capacity)
+    {
+        size_t toAllocate = totalAllocation(capacity);
+        auto data = cast(ubyte[])(parent.allocate(toAllocate));
+        this(data);
+        assert(_blocks * blockSize >= capacity);
+    }
+
+    /// Ditto
+    static if (!is(ParentAllocator == NullAllocator) &&
+        chooseAtRuntime == theBlockSize)
+    this(size_t capacity, uint blockSize)
+    {
+        this._blockSize = blockSize;
+        this(capacity);
+    }
+
+    /**
+    Returns the actual bytes allocated when `n` bytes are requested, i.e.
+    `n.roundUpToMultipleOf(blockSize)`.
+    */
+    pure nothrow @safe @nogc
+    size_t goodAllocSize(size_t n)
+    {
+        return n.roundUpToMultipleOf(blockSize);
+    }
+
+    /**
+    Allocates `s` bytes of memory and returns it, or `null` if memory
+    could not be allocated.
+
+    The `SharedBitmappedBlock` cannot allocate more than the given block size.
+    Allocations are satisfied by searching the first unset bit in the bitmap,
+    and atomically setting it.
+    In rare memory pressure scenarios, the allocation could fail.
+    */
+    @trusted void[] allocate(const size_t s)
+    {
+        import core.atomic : cas, atomicLoad, atomicOp;
+
+        if (s.divideRoundUp(blockSize) != 1)
+            return null;
+
+        // First zero bit position for all values in the 0 - 255 range
+        // for fast lookup
+        static ubyte[] firstZero = [
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+            0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
+            1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2,
+            2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
+            2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
+            4, 4, 4, 4, 4, 4, 4, 4, 5, 5, 5, 5, 6, 6, 7
+        ];
```

Comment:
How about generating this in CTFE?

Reply:
If i declare it static, won't it be inlined in code?